### PR TITLE
feature: add semantic ui theme form components

### PR DESCRIPTION
**Pull request by jacqueswho**
### Reasons for making this change

This is my first pull request and addition to the open-source community. Could you please let me know what else I need to do. My source of inspiration was what was done for rjsf-material-ui project. Because the project was done in typescript, I could not follow the build guide. Could you perhaps assist me with this?

#1537 

### Checklist

* [ ] **I'm updating documentation**
  - [ ] I've [checked the rendering](https://react-jsonschema-form.readthedocs.io/en/latest/#contributing) of the Markdown text I've added
* [ ] **I'm adding or updating code**
  - [ ] I've added and/or updated tests
  - [ ] I've updated [docs](https://react-jsonschema-form.readthedocs.io/) if needed
* [x] **I'm adding a new feature**
  - [ ] I've updated the playground with an example use of the feature

### CODE REVIEW

**Comment by MonkeyAlex** [comment]
Hi @jacqueswho,

I'm unable to build the project.

I've tried adding the following dev dependencies the following:

npm install --save-dev cross-env
npm install --save-dev webpack
npm install --save-dev webpack-cli

yet it still won't compile.

Attached is an output from running **npm install**.
[2020-02-14T10_06_29_651Z-debug.log](https://github.com/rjsf-team/react-jsonschema-form/files/4203984/2020-02-14T10_06_29_651Z-debug.log)

**Comment by jacqueswho** [comment]
Hi @MonkeyAlex 

I do know that is not building, I'm struggling to get it to build, will need assistance on this

**Comment by jacqueswho** [comment]
@MonkeyAlex Hi cant use the latest version because it requires react 16.8 and it's too soon to move over to the latest version of semantic-ui. @epicfaace could you assist me with the build of this repo? Once you can help with the build, I will update the playground and the documentation on the custom props

**Comment by epicfaace** [comment]
@jacqueswho thanks for the PR!

It looks like the travis error says:
```
npm ERR! cipm can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
npm ERR! 
npm ERR! 
npm ERR! Missing: babel-eslint@^10.0.1
npm ERR! Missing: babel-loader@^8.0.6
npm ERR! Missing: chai@^3.3.0
npm ERR! 
npm
```

Thus, can you do `npm install` so that `package-lock.json` is updated, and then commit `package-lock.json`?

**Comment by jacqueswho** [comment]
> @jacqueswho thanks for the PR!
> 
> It looks like the travis error says:
> 
> ```
> npm ERR! cipm can only install packages when your package.json and package-lock.json or npm-shrinkwrap.json are in sync. Please update your lock file with `npm install` before continuing.
> npm ERR! 
> npm ERR! 
> npm ERR! Missing: babel-eslint@^10.0.1
> npm ERR! Missing: babel-loader@^8.0.6
> npm ERR! Missing: chai@^3.3.0
> npm ERR! 
> npm
> ```
> 
> Thus, can you do `npm install` so that `package-lock.json` is updated, and then commit `package-lock.json`?

Hi @epicfaace @thclark 

Could you assist with the webpack, I am no expert on webpack configurations but still learning

**Comment by hyusetiawan** [comment]
oh man, this would satisfy a lot of requirements for my project. A friendly ping, for this to get merged <3

File: packages/semantic-ui/src/PasswordWidget/PasswordWidget.js [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+    <React.Fragment>
+      <Form.Input
+        id={id}
+        key={id}
+        label={label || schema.title}
+        autoFocus={autofocus}
+        required={required}
+        error={error}
+        disabled={disabled || readonly}
+        name={name}
```
Comment (romgrk) [inline packages/semantic-ui/src/PasswordWidget/PasswordWidget.js]: `name` is not declared.

**Comment by romgrk** [comment]
Some styles used by this PR are strange. I would suggest to not use inverted Headers, and to remove the custom `border: 2px solid black` around some fields. The liberal usage of `Segment` also make some elements have too many borders around them.

[Here](https://github.com/romgrk/react-jsonschema-form/commit/18577ec58b027ad90fd99a9b0829967d4c729238) are some suggested fixes.

File: packages/semantic-ui/src/TextAreaWidget/TextAreaWidget.js [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+      <Form.TextArea
+        id={id}
+        key={id}
+        label={label || schema.title}
+        placeholder={placeholder}
+        error={error}
+        autoFocus={autofocus}
+        required={required}
+        disabled={disabled || readonly}
+        name={name}
```
Comment (romgrk) [inline packages/semantic-ui/src/TextAreaWidget/TextAreaWidget.js]: `name` is not declared.

File: packages/semantic-ui/src/RangeWidget/RangeWidget.js [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+    <React.Fragment>
+      <Form.Input
+        id={id}
+        key={id}
+        {...sliderProps}
+        error={error}
+        label={label || schema.title}
+        required={required}
+        disabled={disabled || readonly}
+        name={name}
```
Comment (romgrk) [inline packages/semantic-ui/src/RangeWidget/RangeWidget.js]: `name` is not declared.

File: packages/semantic-ui/src/TextWidget/TextWidget.js [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+    <React.Fragment>
+      <Form.Input
+        key={id}
+        id={id}
+        required={required}
+        error={error}
+        label={label || schema.title}
+        autoFocus={autofocus}
+        disabled={disabled || readonly}
+        name={name}
```
Comment (romgrk) [inline packages/semantic-ui/src/TextWidget/TextWidget.js]: not declared

File: packages/semantic-ui/src/UpDownWidget/UpDownWidget.js [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+      <Form.Input
+        id={id}
+        key={id}
+        autoFocus={autofocus}
+        required={required}
+        error={error}
+        type="number"
+        label={label}
+        disabled={disabled || readonly}
+        name={name}
```
Comment (romgrk) [inline packages/semantic-ui/src/UpDownWidget/UpDownWidget.js]: not declared

**Comment by jacqueswho** [comment]
> Some styles used by this PR are strange. I would suggest to not use inverted Headers, and to remove the custom `border: 2px solid black` around some fields. The liberal usage of `Segment` also make some elements have too many borders around them. `Button.Group` can be used when there are various adjacent `Button`s.
> 
> [Here](https://github.com/romgrk/react-jsonschema-form/commit/18577ec58b027ad90fd99a9b0829967d4c729238) are some suggested fixes.

Sorry for the late response. Will merge with your fork soon.

**Comment by jacqueswho** [comment]
@romgrk I have merged your changes except, I left the border: 2px solid black. Going to make this an option to show nesting. As it's very easy to the eye to see the nesting with the border than without it

File: packages/semantic-ui/package.json [hunk 6]
```diff
@@ -0,0 +1,126 @@
...
+    "lib"
+  ],
+  "engineStrict": false,
+  "engines": {
+    "npm": ">=2.14.7",
+    "node": ">=6"
+  },
+  "scripts": {
+    "build:lib": "rimraf lib && cross-env NODE_ENV=production babel -d lib/ src/",
+    "build:dist": "rimraf dist && cross-env NODE_ENV=production",
```
Comment (zlove) [inline packages/semantic-ui/package.json]:
```suggestion
    "build:dist": "rimraf dist && cross-env NODE_ENV=production babel -d dist/ src/",
```

Doesn't this script line need something more, like a `babel` command.

**Comment by scottblueink** [comment]
@jacqueswho - thanks for the great work on this. Regarding showNesting, my preference would be to have this false by default. The 2px border does make the nesting very apparent, but it is also somewhat jarring and not very consistent with the default semantic ui styles.

**Comment by zlove** [comment]
What's the correct way to pass custom options to the ArrayTemplateField? It looks like ArrayTemplateField in this semantic-ui theme is expecting that `showNesting` could be overridden through `options` that are passed to the ArrayTemplateField component. But any props I set in `ui:options` in the uiSchema did not show up in `options`.  See this screenshot:

DevTools_-_localhost_5000_dashboard_signers_12192

It looks like the `ui:options` is the [documented behavior](https://react-jsonschema-form.readthedocs.io/en/latest/advanced-customization/#custom-widget-options) to override options for Widgets. Should that work for Template components too?

FYI, the only way I could figure out to successfully pass an `option` to ArrayFieldTemplate was as follows:

```javascript
<Form
    schema={schema}
    uiSchema={uiSchema}
    ArrayFieldTemplate={props => <ArrayFieldTemplate {...props} options={{ showNesting: true }} />}
/>
```

**Comment by jacqueswho** [comment]
> What's the correct way to pass custom options to the ArrayTemplateField? It looks like ArrayTemplateField in this semantic-ui theme is expecting that `showNesting` could be overridden through `options` that are passed to the ArrayTemplateField component. But any props I set in `ui:options` in the uiSchema did not show up in `options`. See this screenshot:
> 
> DevTools_-_localhost_5000_dashboard_signers_12192
> 
> It looks like the `ui:options` is the [documented behavior](https://react-jsonschema-form.readthedocs.io/en/latest/advanced-customization/#custom-widget-options) to override options for Widgets. Should that work for Template components too?
> 
> The only way I could figure out to successfully pass an `option` to ArrayFieldTemplate was as follows:
> 
> ```js
> <Form
>     schema={schema}
>     uiSchema={uiSchema}
>     ArrayFieldTemplate={props => <ArrayFieldTemplate {...props} options={{ showNesting: true }} />}
> />
> ```

Hi @zlove I assumed that ui:options would be passed to the template. I'm waiting for this request to be merged in so I can update the playground to include semantic ui options. Apologies for the oversight. But I could change it to look at ui:options in the ui:schema?

**Comment by zlove** [comment]
@jacqueswho I'm not sure if this library has settled on what is the "correct" way to pass options / props to a component like the ArrayTemplateField. It looks like there is ongoing discussion related to this question [in issue 1647](https://github.com/rjsf-team/react-jsonschema-form/issues/1647).

**Comment by zlove** [comment]
@jacqueswho I created a pull request on your pull request feature branch with quite a few additions / changes. Please [take a look](https://github.com/jacqueswho/react-jsonschema-form/pull/1).

**Comment by epicfaace** [comment]
@jacqueswho thanks for your help! Now that we have the infrastructure for tests / etc. for material-ui ready, we should be able to merge this in soon.

Could you add some simple snapshot tests such as the ones in material-ui (https://github.com/rjsf-team/react-jsonschema-form/tree/master/packages/material-ui/test)?

**Comment by jacqueswho** [comment]
> @jacqueswho thanks for your help! Now that we have the infrastructure for tests / etc. for material-ui ready, we should be able to merge this in soon.
> 
> Could you add some simple snapshot tests such as the ones in material-ui (https://github.com/rjsf-team/react-jsonschema-form/tree/master/packages/material-ui/test)?

@epicfaace I did try to make the same tests, but further complications come in when you reference outside modules. Testing is not my strong suit and honestly could not get this to work. Could you please just merge this in for now and see if you can get the tests to work.

**Comment by jacqueswho** [comment]
@epicfaace Is there a section where you add documentation on different themes. There are some extra semantic prop options that need to be documented on how. Either in the playground for semantic, we show all options to pass, or mention it in your documentation. Also do you publish the "@rjsf/semantic-ui" package to npm?

**Comment by epicfaace** [comment]
Description appears before title, when it should appear after title:

bootstrap:
image


semantic ui:
image

**Comment by epicfaace** [comment]
Checkboxes -- I think they should not be inline by default?

bootstrap:
image


semantic-ui:
image

**Comment by epicfaace** [comment]
"Numbers" example -- can we use semantic ui sliders here?

image


See https://semantic-ui.com/modules/checkbox.html#slider

**Comment by epicfaace** [comment]
"Errors" tab -- I believe "should be boolean" should be showing up a bit below the checkbox?

image

**Comment by epicfaace** [comment]
Alternative date time components don't look right

image

**Comment by epicfaace** [comment]
Need some extra space underneath the dropdown in the anyof field:

image

**Comment by epicfaace** [comment]
For the password field, rather than having the label be part of the textbox, it might be best to be consistent to show the label above the textbox:

current behavior
image

ideal behavior ([codepen](https://codesandbox.io/s/semantic-ui-example-34l4w?file=/example.js))
image

**Comment by rlingineni** [comment]
I think this is a great idea! I had the same requirement to use Semantic UI. However, my project doesn't use the `semantic-react` project. And I didn't want to add it as a dependency. I just use the `semantic.css` directly. 

If anyone else is in the same boat - here's a [super simple gist](https://gist.github.com/rlingineni/bf8f615d83f1724959fe84e395df0f8e)  that you can use to get started with using the builtin template props using semantic-ui. 

image

**Comment by jacqueswho** [comment]
> "Numbers" example -- can we use semantic ui sliders here?
> 
> image
> 
> See https://semantic-ui.com/modules/checkbox.html#slider

please be aware that is not react-semantic documentation
[chceckbox](https://react.semantic-ui.com/modules/checkbox/)

**Comment by jacqueswho** [comment]
> Description appears before title, when it should appear after title:
> 
> bootstrap:
> image
> 
> semantic ui:
> image

@epicfaace this at the moment will not be possible as the description is not one of the props for the individual fields. In order for this to work, you would need to make a small change to pass the `ui:description` down as a prop. MaterialUI has the same problem, if you see below their code, the children are above the `rawDescription `. I have done the same for semantic, but if this was passed down to the fields, we could use the description field for each custom Widget field, if present and show above the input. Let me know on the thoughts


```JSX
const FieldTemplate = ({
  id,
  children,
  displayLabel,
  rawErrors = [],
  rawHelp,
  rawDescription,
}: FieldTemplateProps) => {
  return (
    <FormControl fullWidth={true} error={rawErrors.length ? true : false}>
      {children}
      {displayLabel && rawDescription ? (
        <Typography variant="caption" color="textSecondary">
          {rawDescription}
        </Typography>
      ) : null}
      {rawErrors.length > 0 && (
        <List dense={true}>
          {rawErrors.map((error, i: number) => {
            return (
              <ListItem key={i}>
                <FormHelperText id={id}>- {error}</FormHelperText>
              </ListItem>
            );
          })}
        </List>
      )}
      {rawHelp && <FormHelperText id={id}>{rawHelp}</FormHelperText>}
    </FormControl>
  );
};

export default FieldTemplate;

```

**Comment by jacqueswho** [comment]
Hi @epicfaace 

sorry for the long wait. have been really busy with work and deadlines. I have resolved and fixed most of the issues you have described. Except for the description issue which I have commented on.

Please see the latest changes, I have also updated the documentation with the custom semantic options. I think you should have a section in the future for other themes, as there will be custom theme props that need to be documented somewhere. Just the tests now need to be done. I am no expert on react tests, but if some one could start them or show some examples for the semantic package, I could then follow by example.

**Comment by zlove** [comment]
Nice work @jacqueswho. 

Would lifting labels back up to the FieldTemplate component be a solution for the placement of the description? I prefer that because then if a user of the library wants to customize the handling of labels / descriptions / help they can do it in one component. Otherwise, every widget needs to be modified. Plus, this seems more DRY than repeatedly handling the label / description in every widget.

**Comment by chiradeep** [comment]
When an input does not match the format (or is required and left empty), can we add the `error` className so that it becomes `<div class="ui fluid input error">`. This appears to be the behavior in bootstrap
Screen Shot 2020-06-11 at 10 22 05 AM

**Comment by jacqueswho** [comment]
> When an input does not match the format (or is required and left empty), can we add the `error` className so that it becomes `<div class="ui fluid input error">`. This appears to be the behavior in bootstrap
> Screen Shot 2020-06-11 at 10 22 05 AM

This is a semantic UI theme, using semantic UI components, nothing to do with bootstrap, the bootstrap theme needs its error class for it to display correctly.

**Comment by MonkeyAlex** [comment]
It's really great to see this PR finally being merged into master.
Tremendous effort by all contributors.

I can't seem to find the package "@rjsf/semantic-ui" in NPM.

Without it, the import line would fail
`import Form from "@rjsf/semantic-ui";`

Am I missing something?

**Comment by epicfaace** [comment]
@MonkeyAlex It will be released when the next version is published to npm (probably within the next week).

**Comment by gagandeep** [comment]
Waiting for it. Nice work whole rjsf team.

**Pull request merged**